Add api-loadbalancer option to cli set public or internal loadbalancer

mikesplain opened the pull request:
I found the updates proposed in https://github.com/kubernetes/kops/pull/2456 very helpful and thought it'd be helpful to set this from the command line. I image we're trying to limit the number of flags on `kops cluster create` but considering we allow changing the topology, changing the api server loadbalancer to public or internal seems to make things much easier for people like me who only expose ingress endpoints publicly.  

I'm not super familiar with the code so I'm open to any changes or naming ideas but hope others can find this as useful as I!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2559)
<!-- Reviewable:end -->

ottoyiu commented on cmd/kops/create_cluster.go:
```diff
@@ -760,19 +765,34 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
+			case "internal":
+				cluster.Spec.API.LoadBalancer.Type = api.LoadBalancerTypeInternal
+			default:
+				return fmt.Errorf("unkown api-loadbalancer type: %q", c.APILoadbalancer)
```
Noticed a typo here :)

mikesplain replied:
Good eye, I'll push a fix!

ottoyiu commented on cmd/kops/create_cluster.go:
```diff
@@ -760,19 +765,34 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 	if cluster.Spec.API.IsEmpty() {
-		switch cluster.Spec.Topology.Masters {
-		case api.TopologyPublic:
-			cluster.Spec.API.DNS = &api.DNSAccessSpec{}
```
So by specifying the Loadbalancer flag, DNS access for the API is no longer available?

WillemMali replied:
We should probably guard against usage of conflicting flags and document the conflict.

mikesplain replied:
I guess that's a good point, however by setting Topology to private this was already the case. Any thoughts on the best way to handle this?

justinsb replied:
I think we didn't expose the api options through flags previously.

Public topologies would default to DNS, Private topologies would default to ELB.

I think if we have conflicting flags - i.e. if we can't even populate the cluster consistently - we should return error immediately.  But if we can populate the cluster, we should use validation to verify the flags make sense, so that no matter how the cluster is populated we perform the same validation.

So in this case, I don't think we have a conflict ... yet.

chrislovecnm commented:
@k8s-bot ok to test

mikesplain commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[cmd/kops/create_cluster.go, line 793 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/2559#-KjxoVMSb-j0on6HDH2n-r1-793:-KkBi7POsfdHOBAOWuBg:b-896fix) ([raw file](https://github.com/kubernetes/kops/blob/5df9478af42fb65fe8560bf95f97b06ff05073b9/cmd/kops/create_cluster.go#L793)):*
<details><summary><i>Previously, mikesplain (Mike Splain) wrote…</i></summary><blockquote>

Good eye, I'll push a fix!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/2559)*
<!-- Sent from Reviewable.io -->

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -270,6 +273,8 @@ func NewCmdCreateCluster(f *util.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().StringVar(&options.MasterTenancy, "master-tenancy", options.MasterTenancy, "The tenancy of the master group on AWS. Can either be default or dedicated.")
 	cmd.Flags().StringVar(&options.NodeTenancy, "node-tenancy", options.NodeTenancy, "The tenancy of the node group on AWS. Can be either default or dedicated.")
 
+	cmd.Flags().StringVar(&options.APILoadbalancer, "api-loadbalancer", options.APILoadbalancer, "Sets the API loadbalancer to either 'public' or 'internal'")
```
The field here is `API.LoadBalancer.Type`.  How about we make the flag `api-loadbalancer-type`?  It makes the flag more self documenting, and has a nice symmetry IMO.

(And then it doesn't really matter, because it is not user facing, but we could name the option `APILoadbalancerType` or `APILoadBalancerType`)

mikesplain replied:
Great point! Pushed an update to address this.

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -760,19 +765,34 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
+			cluster.Spec.API.LoadBalancer.Type = api.LoadBalancerTypePublic
+		} else {
+			switch c.APILoadbalancer {
+			case "public":
```
I think you could also do `case "", "public":` here, to get rid of the if statement and (minor) code duplication.

mikesplain replied:
On second glance you're totally right, much simpler. Thanks!

mikesplain commented:
All comments make sense to me, renamed the flag to `api-loadbalancer-type`, cleaned up a few things and renamed `APILoadbalancer` to `APILoadBalancerType`.

justinsb commented:
Going to merge this.  If there's a flag conflict I've overlooked, we can handle that case later :-)

mikesplain commented:
Great thanks!
